File: 2. Medium/200. Number of Islands/DfsSolution.java
class Solution {
    private final int[][] directions = new int[][] { { 0, -1 }, { 1, 0 }, { 0, 1 }, { -1, 0 } };
    private final char LAND = '1';
    private int height;
    private int width;

    public int numIslands(char[][] grid) {
        height = grid.length;
        width = grid[0].length;
        boolean[][] visited = new boolean[height][width];

        int islands = 0;
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                if (visited[i][j] || grid[i][j] != LAND)
                    continue;

                visited[i][j] = true;
                islands++;
                dfs(new int[] { i, j }, grid, visited);
            }
        }

        return islands;
    }

    private void dfs(int[] cur, char[][] e, boolean[][] v) {
        for (int[] dir : directions) {
            int row = cur[0] + dir[0];
            int col = cur[1] + dir[1];

            if (row < 0 || col < 0 || row >= height || col >= width || v[row][col])
                continue;

            v[row][col] = true;
            if (e[row][col] == LAND)
                dfs(new int[] { row, col }, e, v);
        }
    }
}

File: 2. Medium/200. Number of Islands/dfs-solution.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
  const directions = [
    { x: 0, y: 1 },
    { x: 1, y: 0 },
    { x: 0, y: -1 },
    { x: -1, y: 0 },
  ];

  let islands = 0;

  for (let row = 0; row < grid.length; row++) {
    for (let col = 0; col < grid[row].length; col++) {
      if (grid[row][col] != "1") continue;
      islands += dfs(row, col) > 0 ? 1 : 0;
    }
  }

  return islands;

  function dfs(row, col) {
    if (
      row < 0 ||
      row >= grid.length ||
      col < 0 ||
      col >= grid[row].length ||
      grid[row][col] != "1"
    ) return 0;

    grid[row][col] = undefined;
    return directions.reduce((acc, dir) => acc + dfs(row + dir.x, col + dir.y), 1);
  }
};
